### DISCUSSION TITLE
How to create a valid WebSocket close frame (server to client) using only ArrayBuffer, DataView, TypedArray, or Node.js Buffer?

### DISCUSSION THREAD

**Discussion by guest271314** [socketio/socket.io#5005] [Q&A] [answered]
I created a WebSocket server that runs from the browser https://github.com/guest271314/direct-sockets-http-ws-server. I used code in a gist to parse and create frames. The gist code does not contain creating a valid close frame. Right now I'm using AbortController to about the WHATWG Stream.

How to create a valid WebSocket close frame (server to client) using only ArrayBuffer, DataView, TypedArray, or Node.js Buffer?

**Comment by darrachequesne** [comment, marked as answer]
Hi! I think you can take a look at the `ws` implementation: https://github.com/websockets/ws/blob/96c9b3deddf56cacb2d756aaa918071e03cdbc42/lib/sender.js#L146-L198

**Reply by guest271314** [reply]
I looked at that file before asking. I tried something like 

```
// const Buffer.allocUnsafe(2);
// target[0] = 0x80;
// target.writeUInt16BE(0x88, 0);
``` 

Didn't cleanly close. I found this https://stackoverflow.com/a/17177146 and used `New Uint8Array([136, 0])` (or `new Uint8Array([0x88, 0x00])`) which works as expected for normal close without payload. A table with all of the close options reflected in a table in TypedArray form would be helpful. Thanks.
